File: java/revenj-servlet/src/main/java/org/revenj/server/commands/SubmitEvent.java
package org.revenj.server.commands;

import org.revenj.patterns.*;
import org.revenj.security.PermissionManager;
import org.revenj.server.ServerCommand;
import org.revenj.server.CommandResult;
import org.revenj.serialization.Serialization;

import java.io.IOException;
import java.security.Principal;
import java.util.Optional;

public class SubmitEvent implements ServerCommand {

	private final DomainModel domainModel;
	private final PermissionManager permissions;

	public SubmitEvent(
			DomainModel domainModel,
			PermissionManager permissions) {
		this.domainModel = domainModel;
		this.permissions = permissions;
	}

	public static final class Argument<TFormat> {
		public String Name;
		public TFormat Data;
		public Boolean ReturnInstance;

		public Argument(String name, TFormat data, Boolean returnInstance) {
			this.Name = name;
			this.Data = data;
			this.ReturnInstance = returnInstance;
		}

		@SuppressWarnings("unused")
		private Argument() {
		}
	}

	@Override
	public <TInput, TOutput> CommandResult<TOutput> execute(
			ServiceLocator locator,
			Serialization<TInput> input,
			Serialization<TOutput> output,
			TInput data,
			Principal principal) {
		Argument<TInput> arg;
		try {
			arg = input.deserialize(data, Argument.class, data.getClass());
		} catch (IOException e) {
			return CommandResult.badRequest(e.getMessage());
		}
		Optional<Class<?>> manifest = domainModel.find(arg.Name);
		if (!manifest.isPresent()) {
			return CommandResult.badRequest("Unable to find specified domain object: " + arg.Name);
		}
		if (arg.Data == null) {
			return CommandResult.badRequest("Data to submit not specified.");
		}
		if (!DomainEvent.class.isAssignableFrom(manifest.get())) {
			return CommandResult.badRequest("Specified type is not an domain event: " + arg.Name);
		}
		if (!permissions.canAccess(manifest.get(), principal)) {
			return CommandResult.forbidden(arg.Name);
		}
		DomainEvent instance;
		try {
			instance = (DomainEvent) input.deserialize(manifest.get(), arg.Data);
		} catch (IOException e) {
			return CommandResult.badRequest("Error deserializing provided input for: " + arg.Name + ". Reason: " + e.getMessage());
		}
		DomainEventStore store;
		try {
			store = locator.resolve(DomainEventStore.class, manifest.get());
		} catch (ReflectiveOperationException e) {
			return CommandResult.badRequest("Error resolving event store for: " + arg.Name + ". Reason: " + e.getMessage());
		}
		String uri;
		try {
			uri = store.submit(instance);
		} catch (IllegalArgumentException e) {
			return new CommandResult<>(null, e.getMessage(), 400);
		}
		try {
			return new CommandResult<>(Boolean.TRUE.equals(arg.ReturnInstance) ? output.serialize(instance) : output.serialize(uri), "Event stored", 201);
		} catch (IOException e) {
			return new CommandResult<>(null, "Error serializing result.", 500);
		}
	}
}

File: csharp/Plugins/Revenj.Plugins.Server.Commands/SubmitEvent.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.Composition;
using System.Diagnostics.Contracts;
using System.Net;
using System.Runtime.Serialization;
using System.Security;
using System.Security.Principal;
using Revenj.Common;
using Revenj.DomainPatterns;
using Revenj.Extensibility;
using Revenj.Processing;
using Revenj.Security;
using Revenj.Serialization;
using Revenj.Utility;

namespace Revenj.Plugins.Server.Commands
{
	[Export(typeof(IServerCommand))]
	[ExportMetadata(Metadata.ClassType, typeof(SubmitEvent))]
	public class SubmitEvent : IServerCommand
	{
		private static Dictionary<Type, ISubmitCommand> Cache = new Dictionary<Type, ISubmitCommand>();

		private readonly IDomainModel DomainModel;
		private readonly IPermissionManager Permissions;

		public SubmitEvent(
			IDomainModel domainModel,
			IPermissionManager permissions)
		{
			Contract.Requires(domainModel != null);
			Contract.Requires(permissions != null);

			this.DomainModel = domainModel;
			this.Permissions = permissions;
		}

		[DataContract(Namespace = "")]
		public class Argument<TFormat>
		{
			[DataMember]
			public string Name;
			[DataMember]
			public TFormat Data;
			[DataMember]
			public bool? ReturnInstance;
		}

		private static TFormat CreateExampleArgument<TFormat>(ISerialization<TFormat> serializer)
		{
			return serializer.Serialize(new Argument<TFormat> { Name = "Module.Event" });
		}

		public ICommandResult<TOutput> Execute<TInput, TOutput>(
			IServiceProvider locator,
			ISerialization<TInput> input,
			ISerialization<TOutput> output,
			IPrincipal principal,
			TInput data)
		{
			var either = CommandResult<TOutput>.Check<Argument<TInput>, TInput>(input, output, data, CreateExampleArgument);
			if (either.Error != null)
				return either.Error;
			var argument = either.Argument;

			var eventType = DomainModel.Find(argument.Name);
			if (eventType == null)
			{
				return
					CommandResult<TOutput>.Fail(
						"Couldn't find event type {0}.".With(argument.Name),
						@"Example argument: 
" + CommandResult<TOutput>.ConvertToString(CreateExampleArgument(output)));
			}
			if (!typeof(IEvent).IsAssignableFrom(eventType))
			{
				return CommandResult<TOutput>.Fail(@"Specified type ({0}) is not an event. 
Please check your arguments.".With(argument.Name), null);
			}
			if (!Permissions.CanAccess(eventType.FullName, principal))
				return CommandResult<TOutput>.Forbidden(argument.Name);
			try
			{
				ISubmitCommand command;
				if (!Cache.TryGetValue(eventType, out command))
				{
					var commandType = typeof(SubmitEventCommand<>).MakeGenericType(eventType);
					command = Activator.CreateInstance(commandType) as ISubmitCommand;
					var newCache = new Dictionary<Type, ISubmitCommand>(Cache);
					newCache[eventType] = command;
					Cache = newCache;
				}
				return command.Submit(input, output, locator, argument.ReturnInstance ?? false, argument.Data);
			}
			catch (ArgumentException ex)
			{
				return CommandResult<TOutput>.Fail(
					ex.Message,
					ex.GetDetailedExplanation() + @"
Example argument: 
" + CommandResult<TOutput>.ConvertToString(CreateExampleArgument(output)));
			}
		}

		private interface ISubmitCommand
		{
			CommandResult<TOutput> Submit<TInput, TOutput>(
				ISerialization<TInput> input,
				ISerialization<TOutput> output,
				IServiceProvider locator,
				bool returnInstance,
				TInput data);
		}

		private class SubmitEventCommand<TEvent> : ISubmitCommand
			where TEvent : IEvent
		{
			public CommandResult<TOutput> Submit<TInput, TOutput>(
				ISerialization<TInput> input,
				ISerialization<TOutput> output,
				IServiceProvider locator,
				bool returnInstance,
				TInput data)
			{
				TEvent domainEvent;
				try
				{
					domainEvent = data != null ? input.Deserialize<TInput, TEvent>(data, locator) : Activator.CreateInstance<TEvent>();
				}
				catch (Exception ex)
				{
					throw new ArgumentException("Error deserializing domain event.", ex);
				}
				var domainStore = locator.Resolve<IEventStore>();
				string uri;
				try
				{
					uri = domainStore.Submit(domainEvent);
				}
				catch (SecurityException) { throw; }
				catch (Exception ex)
				{
					throw new ArgumentException(
						ex.Message,
						data == null
							? new FrameworkException("Error while submitting event: {0}. Data not sent.".With(ex.Message), ex)
							: new FrameworkException(@"Error while submitting event: {0}. Sent data: 
{1}".With(ex.Message, input.Serialize(domainEvent)), ex));
				}
				try
				{
					var command = domainEvent as ICommand;
					Dictionary<string, List<string>> errors;
					if (command != null)
					{
						errors = command.GetValidationErrors();
						if (errors.Count != 0)
							return CommandResult<TOutput>.Return(
								HttpStatusCode.BadRequest,
								output.Serialize(errors),
								"Validation errors");
						return CommandResult<TOutput>.Return(
							HttpStatusCode.OK,
							returnInstance ? output.Serialize(domainEvent) : output.Serialize(uri),
							"Command processed");
					}
					return CommandResult<TOutput>.Return(
						HttpStatusCode.Created,
						returnInstance ? output.Serialize(domainEvent) : output.Serialize(uri),
						"Event stored");
				}
				catch (Exception ex)
				{
					throw new ArgumentException(
						ex.Message,
						new FrameworkException(@"Error serializing result: " + ex.Message, ex));
				}
			}
		}
	}
}
